File: server/src/models/ContentCategory.ts
import {
  Table,
  Model,
  Column,
  IsUUID,
  PrimaryKey,
  Default,
  DataType,
  BelongsTo,
  ForeignKey,
  AllowNull,
} from "sequelize-typescript";
import { Field, ObjectType } from "type-graphql";
import { Content, Category } from ".";

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class ContentCategory extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @Field()
  @IsUUID(4)
  @ForeignKey(() => Category)
  @AllowNull(false)
  @Column
  categoryId: string;

  @BelongsTo(() => Category)
  category: Category;

  @Field()
  @IsUUID(4)
  @ForeignKey(() => Content)
  @AllowNull(false)
  @Column
  contentId: string;

  @BelongsTo(() => Content)
  content: Content;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;
}

File: go-server/internal/models/content_category.go
package models

func init() {
	registerModel(&ContentCategory{})
}

// ContentCategory is the join table for Content<->Category. It does carry
// @Field() on id/categoryId/contentId/createdAt/updatedAt in the Node
// model, but nothing ever returns a ContentCategory (or [ContentCategory])
// over GraphQL — attachCategory/detachCategory return Boolean — so
// type-graphql's schema builder drops it as unreachable, and there's no
// GraphQL type for it here either.
type ContentCategory struct {
	Base

	CategoryID string    `gorm:"not null;index"`
	Category   *Category `gorm:"foreignKey:CategoryID;constraint:OnDelete:CASCADE"`

	ContentID string   `gorm:"not null;index"`
	Content   *Content `gorm:"foreignKey:ContentID;constraint:OnDelete:CASCADE"`
}
